### ISSUE TITLE
Software Template Array Inputs Not Working As Expected

### ISSUE THREAD

**Issue by asummers** [backstage/backstage#7090]
## Current Behavior

Consider the following software template:

```yaml
apiVersion: backstage.io/v1beta2
kind: Template
metadata:
  name: add-addon
  title: Addon
  description: Add addon
spec:
  owner: users
  type: website

  parameters:
    - title: Values
      properties:
       should-be-list:
          title: This is a title and is huge
          description: "this is a description you can't see"
          type: array
          items:
            type: string
  steps:
    - id: fetch-repo
      name: Fetch repo
      action: fetch:plain
      input:
        url: 'https://github.com/asummers/backstage/tree/master'

    - id: fetch-template
      name: Fetch Template
      action: fetch:template
      input:
        url: ./template2
        values:
          list: '{{parameters.should-be-list}}'

    - id: publish
      name: Publish
      action: publish:github:pull-request
      input:
        repoUrl: 'github.com?owner=asummers&repo=backstage'
        title: 'feat: Add addon'
        branchName: 'add-addon'
        description: |
          Add addon

  output:
    links:
      - url: '{{steps.publish.output.remoteUrl}}'
        text: 'Go to PR'

```

The resulting summary appears as expected with two separate elements and not a combined value
![Screen Shot 2021-09-07 at 10 08 01 AM](https://user-images.githubusercontent.com/831648/132370054-e0c910e7-53d8-43df-ad46-080c1000e6b2.png)

But if we consider the template file
```
[
{% set comma = joiner() %}
{% for value in values.list -%}
${{ comma() }}${{ value }}
{%- endfor %}
]

[
{% for value in values.list -%}
${{ value }},
{%- endfor %}
]
```

we get 
```
[

a,.,b,.,c,.,d,:,1,2,3,4,,,w,.,x,.,y,.,z,:,9,8,7,6
]

[
a,.,b,.,c,.,d,:,1,2,3,4,,,w,.,x,.,y,.,z,:,9,8,7,6,
]
```

This means that the array is being converted to a string somewhere under the hood, which results in arrays disappearing.

## Possible Solution

Fix CSS for form and pass along array as is.

## Steps to Reproduce

See ## Current Behavior

## Context

This makes array forms in software templates impossible as nunjucks does not provide split functionality, to correct the comma joining happening prematurely.

## Your Environment

- NodeJS Version (v12): 14
- Operating System and Version (e.g. Ubuntu 14.04): Ubuntu 18.04
- Browser Information: Chrome

**Comment by mtlewis** [comment]
Hi @asummers! Thanks so much for the detailed description of the situation you're facing. The awkwardness of dealing with non-string values in template parameters is something we're aware of - https://github.com/backstage/backstage/issues/6437 includes a proposal to improve this.

The good news is that it's _possible_ today to deal with non-string values, as long as they're convertible to JSON - it's mentioned in the RFC above, but in short, your template files should receive the `list` parameter as an array if you convert the parameter to JSON in your template.yaml file like so:

```diff
   - id: fetch-template
      name: Fetch Template
      action: fetch:template
      input:
        url: ./template2
        values:
-          list: '{{ parameters.should-be-list }}'
+          list: '{{ json parameters.should-be-list }}'
``` 

Hoping this will get you unblocked to use arrays in your templates!

Thanks also for flagging the issue with the styling of array fields. Since this likely requires some code changes to get fixed, can I suggest we create a separate issue to track this part?

**Comment by asummers** [comment]
`json` works for my use case currently. Thanks for that. 

Would you like me to leave this open and create the other issue for the UI, or close this and open another issue for UI, or repurpose this for the UI components?

**Comment by mtlewis** [comment]
Probably the first one if that’s alright with you!

**Comment by asummers** [comment]
Done in #7093.

**Comment by Rugvip** [comment]
According to @benjdlambert this is fixed in the new v3 templates, so gonna close this one :grin:

Please let us know if it's still an issue

**Comment by padupe** [comment]
This solution is currently not working.
Is there any tip on how to use array in inputs?

@rayleshh

**Comment by mustafazeya** [comment]
> ```yaml
> apiVersion: backstage.io/v1beta2
> ```

@Rugvip It still isn't working :(

**Comment by awanlin** [comment]
@mustafazeya have you tried migrating to `v1beta3` as per the guide here: https://backstage.io/docs/features/software-templates/migrating-from-v1beta2-to-v1beta3 ?

**Comment by mustafazeya** [comment]
> @mustafazeya have you tried migrating to `v1beta3` as per the guide here: https://backstage.io/docs/features/software-templates/migrating-from-v1beta2-to-v1beta3 ?

@awanlin Sorry,I think my question was vague and it might not be related to this issue. I am already using using `v1beta3`.

This is the complete question.

When we are using array of strings in our Scaffolder templates we are getting output without space between comma and the next string. 

for Ex:
**Current situation**
$ {{parameter.array}} gives value a,b,c,d,e

**Expected situation**
$ {{parameter.array}} should give value a, b, c, d, e

Do you see if this requirement makes sense or I will have to do some tweaking myself in the middle to achieve this?

**Comment by benjdlambert** [comment]
@mustafazeya can we see your full template.yaml?

**Comment by mustafazeya** [comment]
@benjdlambert 

```yaml

apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: github-onboarding-template
  title: GitHub onboarding template
  description: Template for onboarding a new team to GitHub Organization
  links:
    - title: Talk to team CDX
      url: https://app.slack.com/client/T1C8F1TM2/C02QCKWBJ68
    - title: Connect on Teams
      url: https://teams.microsoft.com/l/channel/19%3a458aece2f8b54e5da6c571db8d54427b%40thread.tacv2/talk-to-team-cdx?groupId=a85d83d0-29e6-4a6b-9439-ed3b2ef4360b&tenantId=5581c9a8-168b-45f0-abd4-d375da99bf9f
spec:
  owner: cdx
  type: service

  parameters:
    - title: Fill below information to onboard your team to GitHub

      required:
        - name
        - displayName
        - owners
      properties:
        name:
          title: Team Name
          type: string
          description: Name of your Team
          pattern: "^[a-z]+(-[a-z]+)*$"
          ui:help: "keep it small,better use acronyms. e.g. cdx,mmb etc. If it's big you can use [-] seperated name like 'customer-profile'"
          ui:autofocus: true
        displayName:
          title: Display Name for Team
          type: string
          description: Display name for your Team.This can have spaces as well as full name e.g. 'Manage My Booking'
        owners:
          title: AD group for managing access
          description: You can either select your existing AD group or create a new one here to manage access to your Github Teams
          type: object
          oneOf:
          - title: Use existing AD group
            properties:
              oldGroup:
                title: Existing AD group
                type: string
                description: Select your existing AD group from the dropdown
                ui:field: OwnerPicker
                ui:options:
                  allowedKinds:
                    - Group
                  allowArbitraryValues: false
            required:
            - oldGroup
          - title: Create new AD group
            properties:
              newGroup:
                title: AD group owners for GitHub Team
                type: array
                description: Email address of the person(s)[could be Techlead/DPO] who should be the `owner(s)` of the new AD group.Click on ADD ITEM button to add more owners.
                items:
                  type: string
                  pattern: ********
                  default: *********
                  ui:help: Email must be within double quotes ""
                uniqueItems: true
                ui:options:
                  orderable: false
            required:
            - newGroup

  steps:
    - id: fetch-docs
      name: Fetch Docs
      action: custom:copy-file-and-append
      input:
        url: https://github.com/****/****/blob/main
        sourceFileName: "teams.tf"
        arrayPosition: -3
        indentationSpace: 4
        newContent: |-
          ${{parameters.name}} = {
            team_description  = "GitHub team for ${{parameters.displayName}}"
            ad_group_owners   = [${{parameters.owners.newGroup}}]
            existing_ad_group = "${{parameters.owners.oldGroup}}"
          }

    - id: publishOldGroup
      name: Create PR with existing AD group
      action: publish:github:pull-request
      if: ${{parameters.owners.oldGroup}}
      input:
        repoUrl: "github.com?repo=*****&owner=****"
        title: "Github Onboarding for ${{parameters.displayName}}"
        branchName: gh-onboarding-${{parameters.name}}
        description: |
          # Github Onboarding: ${{parameters.displayName}}
          Requested by ${{ user.entity.spec.profile.displayName }}[${{ user.entity.spec.profile.email }}]
          Github Team: ${{parameters.name}}
          AD Group: ${{parameters.owners.oldGroup }}

    - id: publishNewGroup
      name: Create PR with New AD group
      action: publish:github:pull-request
      if: ${{parameters.owners.newGroup}}
      input:
        repoUrl: "github.com?repo=*****&owner=******"
        title: "Github Onboarding for ${{parameters.displayName}}"
        branchName: gh-onboarding-${{parameters.name}}
        description: |
          # Github Onboarding: ${{parameters.displayName}}
          Requested by ${{ user.entity.spec.profile.displayName }}[${{ user.entity.spec.profile.email }}]
          Github Team: ${{parameters.name}}
          AD Group: github-${{parameters.name}}


  output:
    links:
      - title: Go to PR
        url: ${{steps.publishOldGroup.output.remoteUrl }}

      - title: Go to PR
        url: ${{steps.publishNewGroup.output.remoteUrl}}

```

**Comment by mustafazeya** [comment]
`custom:copy-file-and-append` action copies everything from the location and then append some content to the selected file.

**Comment by benjdlambert** [comment]
@mustafazeya and what is the problem exactly? Where's the array not an array? My guess is that it's stringifying the array in nunjucks. But that's just a guess.

**Comment by mustafazeya** [comment]
@benjdlambert My question is following.
If you check my template file, I have an array parameter `owners` and then I am using  a parameter `ad_group_owners   = [${{parameters.owners.newGroup}}]` in `fetch-docs` step.

Currently` [${{parameters.owners.newGroup}}]` gives me value `["abc@google.com","def@google.com","pqr@google.com"]` but what I want is space between each element of array i.e, `["abc@google.com", "def@google.com", "pqr@google.com"]` . 

We are using `pre-commit` in our repos and pre-commit fails because of this which leads to failure of our automatic merge of PR.

**Comment by benjdlambert** [comment]
@mustafazeya hmm, something that springs to mind is that you could run [`terraform fmt`](https://developer.hashicorp.com/terraform/cli/commands/fmt) on the files that you create before pushing and commiting, It's normally good practice for making sure that your autogenerated files have the correct formatting and spacing.

I wonder if you could perhaps do something like this: `ad_group_owners   = [${{parameters.owners.newGroup | join(", ") }}]`?

**Comment by mustafazeya** [comment]
@benjdlambert Thanks ! I can try the explicit join function

https://mozilla.github.io/nunjucks/templating.html#join

**Comment by rmanik54** [comment]
> @mustafazeya hmm, something that springs to mind is that you could run [`terraform fmt`](https://developer.hashicorp.com/terraform/cli/commands/fmt) on the files that you create before pushing and commiting, It's normally good practice for making sure that your autogenerated files have the correct formatting and spacing.
> 
> I wonder if you could perhaps do something like this:
> 
> ```
> ad_group_owners   = [${{parameters.owners.newGroup | join(", ") }}]`
> ```
> 
> To be clear, this problem is not to do with the arrays not being supported like the original issue, it's just the serialization of them into a format the `terraform` is happy with, which is why I think it would be better to run `fmt` on after the generated files are rendered.

Since we need to do Terraform fmt, is there any scaffolder template action to run Terraform commands like fmt?
